gridmet_aggregation_PRMS.py

```diff
@@ -0,0 +1,105 @@
+    :param str gpkg_layer: Provide gpkg_layer name if shp is path to gpkg
+    :return: Geodataframe of the original ncdf file with geospatial information (e.g. geom)
+    """
+
+    ## Read in
+    xr_mapped_df = xr.open_dataset(ncdf_path, decode_times=True).to_dataframe().reset_index()
+    if isinstance(shp, gpd.geodataframe.GeoDataFrame):
+        gdf = shp
+    elif os.path.isfile(shp):
+        gdf_prms_path_edited = shp
+        gdf = gpd.read_file(gdf_prms_path_edited, layer=gpkg_layer)
+    else:
+        print('shp must be path to geospatial file or a geodataframe')
+
+    ## Merge ncdf w/ shapefile (the shpfile has area info) & convert to GeoDataFrame
+    gridmet_drb_df = xr_mapped_df.merge(gdf, how ='left', left_on = left_on, right_index = right_on_index)
+    gridmet_drb_gdf = gpd.GeoDataFrame(gridmet_drb_df)
+
+    return gridmet_drb_gdf
+
+# gridmet_prms_area_avg_agg() function aggregates to groupby_cols scale and outputs the aggregated dataset
+def gridmet_prms_area_avg_agg(df, groupby_cols, val_colnames, wgt_col, output_path = None):
+    """
+    :param pd.DataFrame or gpd.GeoDataFrame df:
+    :param  list groupby_cols: group by cols of df
+    :param list val_colnames: list of columns to perform area-weighted average calculation on
+    :param str wgt_col: weight column
+    :param str output_path: if not None, path to save output df dataset in local dir.
+    :return: df with gridmet metrics per PRMS_segid
+    """
+
+    ## Weighted average calc = sum((val[1] * weight[1]), ... , (val[n] * weight[n]) ) / sum(weight[1],...,weight[n])
+
+    start = time.perf_counter()
+
+    df = df.copy()
+    ## 1. Calc value x weight col
+    for col in val_colnames:
+           df[f'{col}_x_area'] = df[col]*df[wgt_col]
+
+    ## 2. Create group by, by summing new value cols weighted by area
+    df_grouped = df.groupby(groupby_cols).agg(area_m2_sum = (wgt_col, 'sum'),
+                                              tmmx_wgtd_sum = ('tmmx_x_area', 'sum'),
+                                              tmmn_wgtd_sum = ('tmmn_x_area', 'sum'),
+                                              pr_wgtd_sum = ('pr_x_area', 'sum'),
+                                              srad_wgtd_sum = ('srad_x_area', 'sum'),
+                                              vs_wgtd_sum = ('vs_x_area', 'sum'),
+                                              rmax_wgtd_sum = ('rmax_x_area', 'sum'),
+                                              rmin_wgtd_sum = ('rmin_x_area', 'sum'),
+                                              sph_wgtd_sum = ('sph_x_area', 'sum')
+                                              )
+    ## 3. Get weighted average
+    for col in val_colnames:
+        df_grouped[f'{col}_wgtd_avg'] = df_grouped[f'{col}_wgtd_sum'] / df_grouped['area_m2_sum']
+
+    # remove interim col
+    df_final = df_grouped.loc[:, ~(df_grouped.columns.str.contains('wgtd_sum'))]
```

Comment:
I did some playing around with this, and came up with a simplified approach. My main motivation for this change is to leave the variable names un-changed. I think it would be easiest for another project to use the product of this if the variable names are the same. For example, `tmmx` instead of `tmmx_wgtd_sum`. 

What do you think of that idea? 

If you are on-board, this code does the equivalent of what you were doing, but also leaves out the `hru_area_m2` column from the end product (I didn't think that was really needed).

```suggestion
    df.loc[:, val_colnames] = df[val_colnames].values * df[[wgt_col]].values

    ## 2. Create group by, by summing new value cols weighted by area
    df_grouped = df.groupby(groupby_cols).sum()

    ## 3. Get weighted average
    df_grouped.loc[:, val_colnames] = df_grouped[val_colnames].values / df_grouped[[wgt_col]].values

    df_final = df_grouped[val_colnames]
```